(feat) Set up Emailing service to get worker confirmations

joshuaNewman10 opened the pull request.

joshuaNewman10 committed:
(feat) Set up Emailing service to get worker confirmations

kyleaowen commented on client/js/actions/server-action-creator.js:
```diff
@@ -1,5 +1,11 @@
 var AppDispatcher = require('../dispatcher/dispatcher');
 var AppConstants = require('../constants/constants');
+var DayStore = require('../stores/day-store');
```
Day store is unused in here

joshuaNewman10 replied:
Good catch, forgot to take it out after the refactor

kyleaowen commented on server/db/models/User.js:
```diff
@@ -43,6 +43,15 @@ module.exports = function (sequelize, DataTypes) {
         });
       },
 
+      getEmployeeEmail: function (id) {
```
we should probably just add a parameter to getEmployeeInfo which accepts an array of strings with properties on them. that can be in a refactor later on

joshuaNewman10 replied:
Agreed, this seemed like the simplest for now. Will make a comment about this in the User model though so we remebmer

kyleaowen commented on server/email/email-controller.js:
```diff
@@ -0,0 +1,55 @@
+    sendEmails: function (req, res) {
+      var gigId = req.body.gigId;
+
+      Gigs.getGigInfo(gigId).then(function (gigInfo) {
```
this isn't how to promise chain. see below:

```
    var location;
    var gig;

    Gigs.getGigInfo(gigId).then(function (gigInfo) {
      gig = gigInfo;
      return Locations.getLocationInfo(gigInfo.LocationId);
    }).then(function(locationInfo){
      location = locationInfo;
      return UserGigs.getUserGigs(gigId)
    }).then(function(userIds){
      return models.sequelize.Promise.map(userIds, function(userId){
        return Users.getEmployeeEmail(userId.UserId);
      });
    }).then(function(userInfos){
      _.each(userInfos, function(userInfo){
        mandrill.sendEmployeeConfirmationMessage(gigInfo.dataValues, userInfo.dataValues, locationInfo.dataValues);
      }
    }).catch(function(err){
      console.log(err);
    })
```

kyleaowen replied:
also, later on we should refactor the model class methods to accept an array of attributes which it will include / eagerly load

joshuaNewman10 replied:
So in my first version I wasn't using the models.sequelize.Promise.map because we can just shoot the emails off one by one as we get the email addresses from the db. Do you want to refactor to this way for a particular reason? Is it faster/efficient or more readable? Thoughts?

kyleaowen replied:
i see where you're coming from though, but i think that the optimization for sending emails off one by one is a small one since we won't be sending thousands of emails at once.

the models.sequelize.Promise.map way allows us to catch errors the result from getting the users from the db. if we do it the other way, the errors will go unnoticed in our catch block. thoughts?

joshuaNewman10 replied:
Agreed, would rather optimize for our code not erroring out over a efficiency increase unless we find that to be a bottleneck

kyleaowen commented on server/email/email-controller.js:
```diff
@@ -0,0 +1,55 @@
+          UserGigs.getUserGigs(gigId).then(function (userIds) {
+            _.each(userIds, function(userId) {
+              Users.getEmployeeEmail(userId.UserId).then(function (userInfo) {
+                mandrill.sendEmployeeConfirmationMessage(gigInfo.dataValues, userInfo.dataValues, locationInfo.dataValues);
```
you should use the sequelize getters and setter - no need to access the dataValues directly

joshuaNewman10 replied:
Great point, will refactor the instances to use the getters

kyleaowen commented on server/email/email-router.js:
```diff
@@ -0,0 +1,21 @@
+var path = require('path');
+
+module.exports = function(app, io){
+  app.use(express.static(__dirname));
```
is this express.static needed? seems like you serve up the only html from 'confirmation/confirmation.html'

joshuaNewman10 replied:
In an iminent refactor im imagining sending back a different html file if the request failed to go through

kyleaowen commented on server/email/email-router.js:
```diff
@@ -0,0 +1,21 @@
+    res.sendFile(path.join(__dirname, './confirmation.html'));
+  });
+
+  emailRouter.get('/confirmation/response*', emailController.handleResponse);
```
is the star necessary?

joshuaNewman10 replied:
Yes because the url of this get request has information like eventId, userId, confirmation. So I need to grab the whole thing no matter what comes after /confirmation/response

kyleaowen replied:
those things aren't in the query string?

kyleaowen commented on server/email/email-controller.js:
```diff
@@ -0,0 +1,55 @@
+var mandrill = require('./email-service');
```
i think the mandrill variable should be renamed to something like 'emailService' since it is supposed to be an abstraction above mandrill. thoughts?

joshuaNewman10 replied:
Agreed

kyleaowen commented on server/server.js:
```diff
@@ -2,11 +2,12 @@ var express = require('express');
 var app = express();
 var port = 8000;
 var clearDatabase = (process.env.NODE_ENV !== 'production' && process.env.CLEAR_DATABASE === 'true')
+var server = require('http').createServer(app);
```
hey - do you need to use the http library? check out socket.io's docs: http://socket.io/docs/#using-with-express-3/4

joshuaNewman10 replied:
Express deprecated var app = require('express').createServer(); so that actually doesnt work anymore. Ive changed it to the following:

var express = require('express');
var app = express();
var server = require('http').Server(app);
var io = require('socket.io')(server);

kyleaowen replied:
express creates the server for you when you use the listen function. check out the documentation for app.listen : http://expressjs.com/api.html 

actually, i like the code just above - it'll be easy to change it from http to https. just make sure it's 'createServer'

kyleaowen commented:
glad we have emails! please see comments

joshuaNewman10 commented:
Thanks for the review, will make the changes and then resubmit the PR

joshuaNewman10 committed:
(refactor) Refactor to implement PR suggestions

(Refactor) Implement modifications from PR review

(Refactor) Use toJson method instead of building up email info objects

(Bugfix) Add console logs to inspect failing tests

(bugfix) Update mandrill api key settings for production

kyleaowen commented:
hey, didn't realize that this was what you intended to do when passing the dataValues, my bad. i think you should just be able to do gigInfo.toJSON() and that should give you and object with the data properties. i should have read this more carefully earlier
